### ISSUE TITLE
Inaccurate for OpenAI's cookbook example

### ISSUE THREAD

**Issue by iwasrobbed** [hmarr/openai-chat-tokens#5]
Great work on this! I found one edge case where the local token calculation seems to be off by a couple tokens compared to the API response. This example is from [OpenAI's cookbook example](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb):

```typescript
      messages: [
      {
        role: 'system',
        content:
          'You are a helpful, pattern-following assistant that translates corporate jargon into plain English.',
      },
      {
        role: 'system',
        name: 'example_user',
        content: 'New synergies will help drive top-line growth.',
      },
      {
        role: 'system',
        name: 'example_assistant',
        content: 'Things working well together will increase revenue.',
      },
      {
        role: 'system',
        name: 'example_user',
        content:
          "Let's circle back when we have more bandwidth to touch base on opportunities for increased leverage.",
      },
      {
        role: 'system',
        name: 'example_assistant',
        content:
          "Let's talk later when we're less busy about how to do better.",
      },
      {
        role: 'user',
        content:
          "This late pivot means we don't have time to boil the ocean for the client deliverable.",
      },
    ],
    tokens: 129,
```

The result is:
>Expected: 129 (from examples above & validated against OpenAI's API response)
>Received: 121 (from local token calculation)

### Root Cause 
[OpenAI's algorithm](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) shows that you're supposed to add one for each `name` field

If you break it down into individual test cases it seems to show that:

```typescript
{
    messages: [
      {
        role: 'system',
        content:
          'You are a helpful, pattern-following assistant that translates corporate jargon into plain English.',
      },
    ],
    tokens: 25,
  },
  {
    messages: [
      {
        role: 'system',
        name: 'example_user',
        content: 'New synergies will help drive top-line growth.',
      },
    ],
    tokens: 20, // local calc 18
  },
  {
    messages: [
      {
        role: 'system',
        name: 'example_assistant',
        content: 'Things working well together will increase revenue.',
      },
    ],
    tokens: 19, // local calc 17
  },
  {
    messages: [
      {
        role: 'system',
        name: 'example_user',
        content:
          "Let's circle back when we have more bandwidth to touch base on opportunities for increased leverage.",
      },
    ],
    tokens: 28, // local calc 26
  },
  {
    messages: [
      {
        role: 'system',
        name: 'example_assistant',
        content:
          "Let's talk later when we're less busy about how to do better.",
      },
    ],
    tokens: 26, // local calc 24
  },
  {
    messages: [
      {
        role: 'user',
        content:
          "This late pivot means we don't have time to boil the ocean for the client deliverable.",
      },
    ],
    tokens: 26,
  },
```

### Making that change (to add a token for every `message.name`) fixes those tests, but then breaks these ones:
<img width="776" alt="Screenshot 2023-08-09 at 12 56 30 AM" src="https://github.com/hmarr/openai-chat-tokens/assets/317671/f1f04f56-3987-47fb-b7fd-e1fa1fd9bd7e">
<img width="777" alt="Screenshot 2023-08-09 at 12 56 23 AM" src="https://github.com/hmarr/openai-chat-tokens/assets/317671/7a6e1da4-52a7-4347-ac02-18a3d8b0c2fb">
<img width="777" alt="Screenshot 2023-08-09 at 12 56 18 AM" src="https://github.com/hmarr/openai-chat-tokens/assets/317671/ffa68265-f852-4e40-896e-6042c035eea3">

### Final tweaks

So all said, it seems if there's a `function` then you need to `subtract 2`:

```typescript
  if (message.name) {
    tokens += 1 // matches OpenAI's cookbook
  }
  if (message.role === 'function') {
    tokens -= 2
  }
  if (message.function_call) {
    tokens += 3
  }
```

**Comment by hmarr** [comment]
Good find! Thanks for explaining in so much detail. I was aware that the `name` parameter was one of the few cases I hadn't tried out yet so it makes sense that it was off.

Feel free to submit the tests and fix in a PR, or if you'd rather I can easily copy the changes in directly.

**Comment by iwasrobbed** [comment]
Happy to submit a PR as a thanks! Will do tomorrow

**Comment by hmarr** [comment]
Thanks for the PR! Just released in v0.2.5
